File: src/game/powerups/portal/EndLevelPortal.java
package game.powerups.portal;

import game.Bomberman;
import game.data.DataInputOutput;
import game.entity.models.BomberEntity;
import game.entity.models.Coordinates;
import game.level.Level;
import game.level.WorldSelectorLevel;
import game.utils.Paths;

import java.awt.image.BufferedImage;
import java.lang.reflect.InvocationTargetException;

public class EndLevelPortal extends Portal {
    public EndLevelPortal(Coordinates coordinates) {
        super(coordinates);
    }

    @Override
    public BufferedImage getImage() {
        return loadAndSetImage(Paths.getPowerUpsFolder() + "/end_game.gif");
    }

    @Override
    public int getDuration() {
        return 0;
    }

    @Override
    public boolean canPickUp(BomberEntity entity) {
        return Bomberman.getMatch().getEnemiesAlive() <= 0;
    }

    @Override
    protected void doApply(BomberEntity entity) {
        super.doApply(entity);

        Bomberman.getMatch().getCurrentLevel().endLevel();
        DataInputOutput.getInstance().increaseLives();

        try {
            Level level = Bomberman.getMatch().getCurrentLevel();
            Class<? extends Level> nextLevelClass =
                    level.isLastLevelOfWorld()
                            ? WorldSelectorLevel.class
                            : level.getNextLevel();

            Bomberman.startLevel(nextLevelClass.getConstructor().newInstance());
        } catch (InstantiationException | IllegalAccessException | InvocationTargetException |
                 NoSuchMethodException e) {
            e.printStackTrace();
        }
    }

    @Override
    protected void cancel(BomberEntity entity) {
    }
}

File: src/game/powerups/portal/EndLevelPortal.kt
package game.powerups.portal

import game.Bomberman
import game.data.DataInputOutput
import game.entity.EntityTypes
import game.entity.player.BomberEntity
import game.entity.models.Coordinates
import game.level.levels.lobby.WorldSelectorLevel
import game.utils.Paths.powerUpsFolder
import java.awt.image.BufferedImage

class EndLevelPortal(coordinates: Coordinates?) : Portal(coordinates) {
    constructor(id: Long) : this(null) {
        this.id = id
    }

    override fun getImage(): BufferedImage = loadAndSetImage("$powerUpsFolder/end_game.gif")

    override val duration: Int
        get() = 0

    override fun canPickUp(entity: BomberEntity): Boolean {
        return Bomberman.getMatch().enemiesAlive <= 0
    }

    override fun doApply(entity: BomberEntity) {
        super.doApply(entity)

        val match = Bomberman.getMatch()
        val currentLevel = match.currentLevel ?: return

        currentLevel.endLevel()
        DataInputOutput.getInstance().increaseLives()

        try {
            val nextLevelClass = if (currentLevel.info.isLastLevelOfWorld)
                WorldSelectorLevel::class.java
            else currentLevel.info.nextLevel

            Bomberman.startLevel(nextLevelClass!!.getDeclaredConstructor().newInstance(), match.onlineGameHandler)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    override fun cancel(entity: BomberEntity) {}

    override val type: EntityTypes
        get() = EntityTypes.EndLevelPortal
}